File: algorithms/Strings/Medium/LongestPalindromicSubstring/VSDLongestPalindromicSubstring.py


def expandFromCenter(string,left,right):
    global maxlength
    global start
    while left>-1 and right<len(string) and string[left]==string[right]:
        left-=1
        right+=1
    if maxlength<right-left-1:
        maxlength=right-left-1
        start=left+1
global maxlength 
maxlength=0

global start
start=0
def longestPalindrome(string):
    global maxlength
    global start
    for i in range(len(string)):
        expandFromCenter(string,i,i)
        expandFromCenter(string,i,i+1)
    return string[start:start+maxlength]


if __name__=="__main__":
    string="badad"
    print(longestPalindrome(string))

File: algorithms/Strings/Medium/LongestPalindromicSubstring/VSDLongestPalindromicSubstring.cpp


#include<bits/stdc++.h>



int expandFromCenter(std::string& string,int left,int right){
    while(left>-1 && right<string.length() && string[left]==string[right]){
        left--;
        right++;
    }
    return right-left-1;
}


std::string longestPalindrome(std::string& string){
    int start=0,end=0;
    for(int i=0;i<string.length();i++){
        int len1 = expandFromCenter(string,i,i);
        int len2 = expandFromCenter(string,i,i+1);
        int len = std::max(len1,len2);
        if(len > end - start){
            start=i - (len - 1) /2;
            end = i + len /2;
        }

    }
    
    return string.substr(start,end+1);
}


int main(){
    
    std::string string = "badab";
    std::cout<< longestPalindrome(string);
}